File: leetcode/search/1182. Shortest Distance to Target Color/js/main.js
/*


*/

const shortestDistanceColor = (colors, queries) => {
  let map = new Map();
  let res = Array(queries.length).fill(-1);

  for (let i = 1; i <= 3; i++) {
    map.set(i, []);
  }

  for (let i = 0; i < colors.length; i++) {
    map.get(colors[i]).push(i);
  }

  for (let i = 0; i < queries.length; i++) {
    let [index, val] = queries[i];

    if (map.get(val).length) {
      res[i] = search(map.get(val), index);
    }
  }

  return res;

  function search(nums, target) {
    let lo = 0;
    let hi = nums.length - 1;

    while (lo < hi) {
      let mid = lo + Math.floor((hi - lo) / 2);

      if (nums[mid] < target) {
        lo = mid + 1;
      } else {
        hi = mid;
      }
    }

    // Если точного значения нету в массиве, nums[lo] > target
    if (lo > 0 && target - nums[lo - 1] < nums[lo] - target) {
      return target - nums[lo - 1];
    }

    return Math.abs(nums[lo] - target);
  }
};

File: leetcode/search/1182. Shortest Distance to Target Color/py/main.py
import collections

class Solution:
    def shortestDistanceColor(self, colors, queries):
      def search(nums, target):
        lo = 0
        hi = len(nums) - 1

        while lo < hi:
          mid = lo + ((hi - lo) >> 1)

          if nums[mid] < target:
            lo = mid + 1
          else:
            hi = mid
        
        if lo > 0 and target - nums[lo - 1] < nums[lo] - target:
          return target - nums[lo-1]

        return abs(target - nums[lo])

      maps = collections.defaultdict(list)

      for i, v in enumerate(colors):
        maps[v].append(i)
      
      res = []

      for index, val in queries:
        if val in maps:
          res.append(search(maps[val], index))
        else:
          res.append(-1)

      return res